inline/src/lexer/resolver/mod.rs

```diff
@@ -66,209 +91,273 @@ impl TokenMap {
 /// - Every time a token pair is matched, all non-resolved tokens between them are marked as plain
 #[derive(Debug, Clone)]
 pub(crate) struct TokenResolver<'token> {
+    /// Scopes are introduced by brackets (for example text groups).
     curr_scope: usize,
-    interrupted: Vec<Range<usize>>,
-    pub(crate) tokens: Vec<RawToken<'token>>,
+
+    /// Tape that enables expanding *visible* context inside of the TokenIterator.
+    pub(crate) tape: Tape<TokenIter<'token>>,
+
+    /// The index of the `Token` at head of `Tape`. That is, the number of the `Token` that will be
+    /// returned by the resolver, with 0 being the number of the first `Token`.
+    tape_idx: usize,
+
+    /// Mapping of `TokenKind` to indices of `Token`s that are not yet resolved.
+    unresolved: TokenMap,
 }
 
 impl<'token> TokenResolver<'token> {
     pub(crate) fn new(iter: TokenIterator<'token>) -> Self {
-        let mut new = Self {
-            curr_scope: 0,
-            interrupted: Vec::default(),
-            tokens: iter.map(RawToken::new).collect(),
-        };
+        let tape = iter.clone().map(RawToken::new as _).tape();
```

Comment:
```suggestion
        let tape = iter.map(RawToken::new as _).tape();
```

unnecessary clone, because `iter` is not used anymore.